In [95]:
import kagglehub

# Download latest version
path = kagglehub.dataset_download("manishkr1754/cardekho-used-car-data")

print("Path to dataset files:", path)


Using Colab cache for faster access to the 'cardekho-used-car-data' dataset.
Path to dataset files: /kaggle/input/cardekho-used-car-data


In [96]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
import seaborn as sns
import os

In [97]:
os.listdir(path)

['cardekho_dataset.csv']

In [98]:
df=pd.read_csv("/kaggle/input/cardekho-used-car-data/cardekho_dataset.csv")

In [99]:
print("Shape:",df.shape)

Shape: (15411, 14)


In [100]:
df.head()

,Unnamed: 0,car_name,brand,model,vehicle_age,km_driven,seller_type,fuel_type,transmission_type,mileage,engine,max_power,seats,selling_price
0,0,Maruti Alto,Maruti,Alto,9,120000,Individual,Petrol,Manual,19.70,796,46.30,5,120000
1,1,Hyundai Grand,Hyundai,Grand,5,20000,Individual,Petrol,Manual,18.90,1197,82.00,5,550000
2,2,Hyundai i20,Hyundai,i20,11,60000,Individual,Petrol,Manual,17.00,1197,80.00,5,215000
3,3,Maruti Alto,Maruti,Alto,9,37000,Individual,Petrol,Manual,20.92,998,67.10,5,226000
4,4,Ford Ecosport,Ford,Ecosport,6,30000,Dealer,Diesel,Manual,22.77,1498,98.59,5,570000


In [101]:
df.describe()

,Unnamed: 0,vehicle_age,km_driven,mileage,engine,max_power,seats,selling_price
count,15411.000000,15411.000000,1.541100e+04,15411.000000,15411.000000,15411.000000,15411.000000,1.541100e+04
mean,9811.857699,6.036338,5.561648e+04,19.701151,1486.057751,100.588254,5.325482,7.749711e+05
std,5643.418542,3.013291,5.161855e+04,4.171265,521.106696,42.972979,0.807628,8.941284e+05
min,0.000000,0.000000,1.000000e+02,4.000000,793.000000,38.400000,0.000000,4.000000e+04
25%,4906.500000,4.000000,3.000000e+04,17.000000,1197.000000,74.000000,5.000000,3.850000e+05
50%,9872.000000,6.000000,5.000000e+04,19.670000,1248.000000,88.500000,5.000000,5.560000e+05
75%,14668.500000,8.000000,7.000000e+04,22.700000,1582.000000,117.300000,5.000000,8.250000e+05
max,19543.000000,29.000000,3.800000e+06,33.540000,6592.000000,626.000000,9.000000,3.950000e+07


In [102]:
df.columns

Index(['Unnamed: 0', 'car_name', 'brand', 'model', 'vehicle_age', 'km_driven',
       'seller_type', 'fuel_type', 'transmission_type', 'mileage', 'engine',
       'max_power', 'seats', 'selling_price'],
      dtype='object')

In [103]:
numerical_features=df.select_dtypes(include=['int64','float64']).columns
categorical_features=df.select_dtypes(include=['object']).columns
print("Categorical Features")
print(categorical_features.to_list())
print("Numerical Features")
print(numerical_features.to_list())

Categorical Features
['car_name', 'brand', 'model', 'seller_type', 'fuel_type', 'transmission_type']
Numerical Features
['Unnamed: 0', 'vehicle_age', 'km_driven', 'mileage', 'engine', 'max_power', 'seats', 'selling_price']


In [104]:
df.isnull().sum()

,0
Unnamed: 0,0
car_name,0
brand,0
model,0
vehicle_age,0
km_driven,0
seller_type,0
fuel_type,0
transmission_type,0
mileage,0


In [105]:
X=df.drop(['selling_price'],axis=1)
y=df['selling_price']

In [106]:
y

,selling_price
0,120000
1,550000
2,215000
3,226000
4,570000
...,...
15406,250000
15407,925000
15408,425000
15409,1225000


In [107]:
data=df.copy()

In [108]:
data["km_per_year"] = (
    data["km_driven"] /
    data["vehicle_age"].clip(lower=1)
)

In [109]:
data.head()

,Unnamed: 0,car_name,brand,model,vehicle_age,km_driven,seller_type,fuel_type,transmission_type,mileage,engine,max_power,seats,selling_price,km_per_year
0,0,Maruti Alto,Maruti,Alto,9,120000,Individual,Petrol,Manual,19.70,796,46.30,5,120000,13333.333333
1,1,Hyundai Grand,Hyundai,Grand,5,20000,Individual,Petrol,Manual,18.90,1197,82.00,5,550000,4000.000000
2,2,Hyundai i20,Hyundai,i20,11,60000,Individual,Petrol,Manual,17.00,1197,80.00,5,215000,5454.545455
3,3,Maruti Alto,Maruti,Alto,9,37000,Individual,Petrol,Manual,20.92,998,67.10,5,226000,4111.111111
4,4,Ford Ecosport,Ford,Ecosport,6,30000,Dealer,Diesel,Manual,22.77,1498,98.59,5,570000,5000.000000


In [110]:
data["is_automatic"] = (
    data["transmission_type"]
    .str.lower()
    .eq("automatic")
    .astype(int)
)

In [111]:
X = data.drop(columns=["selling_price"])

In [112]:
y = data["selling_price"]

In [113]:
y

,selling_price
0,120000
1,550000
2,215000
3,226000
4,570000
...,...
15406,250000
15407,925000
15408,425000
15409,1225000


In [114]:

X = pd.get_dummies(X, drop_first=True)

In [115]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42
)

In [116]:
import numpy as np
import pandas as pd

from sklearn.linear_model import LinearRegression
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor

from sklearn.metrics import (
    mean_absolute_error,
    mean_squared_error,
    r2_score
)

In [117]:
lr = LinearRegression()

lr.fit(X_train, y_train)

LinearRegression()

In [118]:
y_pred_lr = lr.predict(X_test)

In [119]:
mae_lr = mean_absolute_error(y_test, y_pred_lr)
mse_lr = mean_squared_error(y_test, y_pred_lr)
rmse_lr = np.sqrt(mse_lr)
r2_lr = r2_score(y_test, y_pred_lr)

print("Linear Regression")
print("MAE  :", mae_lr)
print("MSE  :", mse_lr)
print("RMSE :", rmse_lr)
print("R²   :", r2_lr)

Linear Regression
MAE  : 177496.43666635448
MSE  : 150747170421.779
RMSE : 388261.72927778884
R²   : 0.7997465237236747


In [120]:
dt = DecisionTreeRegressor(
    random_state=42
)

In [121]:
dt.fit(X_train, y_train)

DecisionTreeRegressor(random_state=42)

In [122]:
y_pred_dt = dt.predict(X_test)

In [123]:
mae_dt = mean_absolute_error(y_test, y_pred_dt)
mse_dt = mean_squared_error(y_test, y_pred_dt)
rmse_dt = np.sqrt(mse_dt)
r2_dt = r2_score(y_test, y_pred_dt)

print("Decision Tree Regressor")
print("MAE  :", mae_dt)
print("MSE  :", mse_dt)
print("RMSE :", rmse_dt)
print("R²   :", r2_dt)

Decision Tree Regressor
MAE  : 127260.82549464807
MSE  : 94694220326.18391
RMSE : 307724.26021713647
R²   : 0.874207610328353


In [126]:
rf = RandomForestRegressor(
    n_estimators=100,
    random_state=42,
    n_jobs=-1
)

In [127]:
rf.fit(X_train, y_train)

RandomForestRegressor(n_jobs=-1, random_state=42)

In [128]:
y_pred_rf = rf.predict(X_test)

In [129]:
mae_rf = mean_absolute_error(y_test, y_pred_rf)
mse_rf = mean_squared_error(y_test, y_pred_rf)
rmse_rf = np.sqrt(mse_rf)
r2_rf = r2_score(y_test, y_pred_rf)

print("Random Forest Regressor")
print("MAE  :", mae_rf)
print("MSE  :", mse_rf)
print("RMSE :", rmse_rf)
print("R²   :", r2_rf)

Random Forest Regressor
MAE  : 95721.91534219915
MSE  : 47074069226.73532
RMSE : 216965.5945691282
R²   : 0.9374665144377117


In [130]:
results = pd.DataFrame({
    "Model": [
        "Linear Regression",
        "Decision Tree Regressor",
        "Random Forest Regressor"
    ],

    "MAE": [
        mae_lr,
        mae_dt,
        mae_rf
    ],

    "MSE": [
        mse_lr,
        mse_dt,
        mse_rf
    ],

    "RMSE": [
        rmse_lr,
        rmse_dt,
        rmse_rf
    ],

    "R² Score": [
        r2_lr,
        r2_dt,
        r2_rf
    ]
})

results

,Model,MAE,MSE,RMSE,R² Score
0,Linear Regression,177496.436666,1.507472e+11,388261.729278,0.799747
1,Decision Tree Regressor,127260.825495,9.469422e+10,307724.260217,0.874208
2,Random Forest Regressor,95721.915342,4.707407e+10,216965.594569,0.937467


## Model Comparison and Final Selection

Among the three regression models, the **Random Forest Regressor** achieved the best overall performance.

- MAE: 95,721.92
- MSE: 4.7074 × 10¹⁰
- RMSE: 216,965.59
- R² Score: 0.9375

Random Forest achieved the lowest MAE, MSE, and RMSE and the highest R² score. The R² score indicates that the model explains approximately 93.7% of the variation in selling prices in the test data.

Linear Regression achieved an R² score of 0.7997, suggesting that a purely linear model does not fully capture the relationships between the vehicle features and selling price.

The Decision Tree improved performance with an R² score of 0.8742 by capturing nonlinear relationships. However, the Random Forest performed considerably better by combining multiple decision trees, reducing the variance and overfitting associated with a single tree.

Therefore, **Random Forest Regressor was selected as the best-performing model**.

### Future Improvements

1. Perform hyperparameter tuning using GridSearchCV or RandomizedSearchCV.
2. Use cross-validation for more reliable model evaluation.
3. Experiment with boosting algorithms such as Gradient Boosting, XGBoost, LightGBM, or CatBoost.
4. Explore additional feature engineering and investigate price outliers.